File: container/github.py
import functools
import logging
import time

import jwt
import requests

from cachetools.func import ttl_cache
from constants import *

logger = logging.getLogger(__name__)


class GitHubAPIError(Exception):
    """Exception raised for GitHub API errors."""
    def __init__(self, status_code: int, message: str):
        self.status_code = int(status_code)
        self.message = message
        super().__init__(message)


@functools.lru_cache(maxsize=1)
def init_ghapp_private_key_org():
    private_key = jwt.jwk_from_pem(GHAPP_ORG_PRIVATE_KEY.encode('utf-8'))
    assert private_key, "Failed to load private key from GHAPP_ORG_PRIVATE_KEY"
    return private_key


@functools.lru_cache(maxsize=1)
def init_ghapp_private_key_personal():
    private_key = jwt.jwk_from_pem(GHAPP_PERSONAL_PRIVATE_KEY.encode('utf-8'))
    assert private_key, "Failed to load private key from GHAPP_PERSONAL_PRIVATE_KEY"
    return private_key


def generate_jwt(app_id, private_key):
    """Generate a JWT for GitHub App authentication."""
    payload = {
        "iat": int(time.time()),
        "exp": int(time.time()) + (10 * 60),
        "iss": app_id,
    }
    return jwt.JWT().encode(payload, private_key, alg="RS256")


@ttl_cache(maxsize=1024, ttl=60*59) # Authentication Token lifetime is 1 hour
def authenticate_app(installation_id, app_id):
    """Authenticate the app and get an installation token.

    `app_id` selects which app's private key signs the JWT — pass
    `GHAPP_PERSONAL_ID` for user installations, `GHAPP_ORG_ID` for org
    installations.
    """
    if app_id == GHAPP_PERSONAL_ID:
        jwt_token = generate_jwt(GHAPP_PERSONAL_ID, init_ghapp_private_key_personal())
    elif app_id == GHAPP_ORG_ID:
        jwt_token = generate_jwt(GHAPP_ORG_ID, init_ghapp_private_key_org())
    else:
        raise ValueError(f"Unknown app_id: {app_id}")

    headers = {
        "Authorization": f"Bearer {jwt_token}",
        "Accept": "application/vnd.github.v3+json",
    }
    url = f"https://api.github.com/app/installations/{installation_id}/access_tokens"
    response = requests.post(url, headers=headers, json={})

    if response.status_code == 201:
        logger.debug("Obtained installation access token for installation %s", installation_id)
        return response.json().get("token")
    else:
        error = response.json().get("message")
        logger.error("Failed to get installation access token for installation %s: %s", installation_id, error)
        raise GitHubAPIError(response.status_code, f"Failed to get installation access token: {error}")


def get_installation(installation_id, entity_type):
    """Fetch installation metadata for the app matching entity_type.

    Returns the parsed JSON (includes account.type, account.login, app_id).
    """
    assert entity_type is not None
    if entity_type == EntityType.USER:
        jwt_token = generate_jwt(GHAPP_PERSONAL_ID, init_ghapp_private_key_personal())
    else:
        jwt_token = generate_jwt(GHAPP_ORG_ID, init_ghapp_private_key_org())

    headers = {
        "Authorization": f"Bearer {jwt_token}",
        "Accept": "application/vnd.github.v3+json",
    }
    url = f"https://api.github.com/app/installations/{installation_id}"
    response = requests.get(url, headers=headers)

    if response.status_code == 200:
        return response.json()
    else:
        error = response.json().get("message")
        logger.error("Failed to get installation %s: %s", installation_id, error)
        raise GitHubAPIError(response.status_code, f"Failed to get installation: {error}")


def ensure_runner_group(entity_name, token, group_name):
    """Ensure the runner group exists and return its ID."""
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github.v3+json",
    }

    list_url = f"https://api.github.com/orgs/{entity_name}/actions/runner-groups"
    response = requests.get(list_url, headers=headers)
    if response.status_code != 200:
        error = response.json()
        logger.error("Failed to list runner groups for org %s: %s", entity_name, error)
        raise GitHubAPIError(response.status_code, f"Failed to list runner groups: {error}")

    for group in response.json().get("runner_groups", []):
        if group.get("name") == group_name:
            logger.debug("Found existing runner group '%s' (id=%s) for org %s",
                        group_name, group["id"], entity_name)
            return group["id"]

    create_body = {
        "name": group_name,
        "visibility": "all",
        "allows_public_repositories": True,
    }
    response = requests.post(list_url, headers=headers, json=create_body)
    if response.status_code == 201:
        runner_group_id = response.json().get("id")
        logger.debug("Created runner group '%s' (id=%s) for org %s",
                     group_name, runner_group_id, entity_name)
        return runner_group_id
    else:
        error = response.json()
        logger.error("Failed to create runner group '%s' for org %s: %s",
                     group_name, entity_name, error)
        raise GitHubAPIError(response.status_code, f"Failed to create runner group: {error}")


def create_jit_runner_config_org(token, group_id, labels, entity_name, runner_name):
    """Create a JIT runner configuration for a new ephemeral runner (org-scoped)."""
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github.v3+json",
    }
    url = f"https://api.github.com/orgs/{entity_name}/actions/runners/generate-jitconfig"
    body = {
        "name": runner_name,
        "runner_group_id": group_id,
        "labels": labels,
        "work_folder": "../../../work", # /home/runner/actions-runner/cached/X.Y.Z/bin/../ -> /home/runner/work
    }
    response = requests.post(url, headers=headers, json=body)

    if response.status_code == 201:
        jit_config = response.json().get("encoded_jit_config")
        logger.debug("Created JIT runner config for org %s, runner name=%s", entity_name, runner_name)
        return jit_config
    else:
        error = response.json()
        logger.error("Failed to create JIT runner config for org %s: %s", entity_name, error)
        raise GitHubAPIError(response.status_code, f"Failed to create JIT runner config: {error}")


def create_jit_runner_config_repo(token, labels, repo_full_name, runner_name):
    """Create a JIT runner configuration for a new ephemeral runner (repo-scoped, for personal accounts)."""
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github.v3+json",
    }
    url = f"https://api.github.com/repos/{repo_full_name}/actions/runners/generate-jitconfig"
    body = {
        "name": runner_name,
        "runner_group_id": 1,  # default runner group for repos
        "labels": labels,
        "work_folder": "../../../work", # /home/runner/actions-runner/cached/X.Y.Z/bin/../ -> /home/runner/work
    }
    response = requests.post(url, headers=headers, json=body)

    if response.status_code == 201:
        jit_config = response.json().get("encoded_jit_config")
        logger.debug("Created JIT runner config for repo %s, runner name=%s", repo_full_name, runner_name)
        return jit_config
    else:
        error = response.json()
        logger.error("Failed to create JIT runner config for repo %s: %s", repo_full_name, error)
        raise GitHubAPIError(response.status_code, f"Failed to create JIT runner config: {error}")


def _paginated_get(url, token, collection_key):
    """GET a paginated GitHub list endpoint, following the `Link: rel="next"` header.

    Returns the concatenated list under `collection_key` in each response body.
    Raises GitHubAPIError on non-2xx.
    """
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github.v3+json",
    }
    results = []
    next_url = f"{url}?per_page=100"
    while next_url:
        response = requests.get(next_url, headers=headers)
        if response.status_code != 200:
            raise GitHubAPIError(response.status_code, f"GET {next_url}: {response.text}")
        results.extend(response.json().get(collection_key, []))
        next_url = response.links.get("next", {}).get("url")
    return results


def list_runners_org_group(token, entity_name, group_id):
    """List all runners registered under a specific org runner group."""
    url = f"https://api.github.com/orgs/{entity_name}/actions/runner-groups/{group_id}/runners"
    return _paginated_get(url, token, "runners")


def list_runners_repo(token, repo_full_name):
    """List all runners registered under a specific repo."""
    url = f"https://api.github.com/repos/{repo_full_name}/actions/runners"
    return _paginated_get(url, token, "runners")


def delete_runner_org(token, entity_name, runner_id):
    """DELETE an org-scoped runner. 404 is swallowed (already gone)."""
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github.v3+json",
    }
    url = f"https://api.github.com/orgs/{entity_name}/actions/runners/{runner_id}"
    response = requests.delete(url, headers=headers)
    if response.status_code in (204, 404):
        return
    raise GitHubAPIError(response.status_code, f"DELETE {url}: {response.text}")


def delete_runner_repo(token, repo_full_name, runner_id):
    """DELETE a repo-scoped runner. 404 is swallowed."""
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github.v3+json",
    }
    url = f"https://api.github.com/repos/{repo_full_name}/actions/runners/{runner_id}"
    response = requests.delete(url, headers=headers)
    if response.status_code in (204, 404):
        return
    raise GitHubAPIError(response.status_code, f"DELETE {url}: {response.text}")


def get_job_info(repo_full_name, job_id, token):
    """Get the effective status of a workflow job from GitHub API.

    GitHub can return status="in_progress" with conclusion="cancelled" (or other
    terminal conclusions). When a conclusion is present, the job is effectively
    completed regardless of the status field.
    """
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github.v3+json",
    }
    url = f"https://api.github.com/repos/{repo_full_name}/actions/jobs/{job_id}"
    response = requests.get(url, headers=headers)

    if response.status_code == 200:
        return response.json()
    else:
        logger.error("Failed to get job status for %s job %s: %s", repo_full_name, job_id, response.status_code)
        raise GitHubAPIError(response.status_code, f"Failed to get job status: {response.text}")

File: container/internal/github.go
package internal

import (
	"bytes"
	"context"
	"crypto/rsa"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

// HTTPDoer is the minimal interface ghClient needs from *http.Client.
// Tests pass an httptest.Server-backed *http.Client; production uses http.DefaultClient.
type HTTPDoer interface {
	Do(*http.Request) (*http.Response, error)
}

// GHClient implements GitHubClient over the live GitHub REST API.
// Installation tokens are issued by GitHub with a 1-hour lifetime; we cache
// them for 59 minutes (= 1 hour minus clock skew safety) keyed by
// (installation_id, app_id).
type GHClient struct {
	HTTP      HTTPDoer
	BaseURL   string // defaults to https://api.github.com; tests override
	OrgAppID  int64
	PersAppID int64
	OrgKey    *rsa.PrivateKey
	PersKey   *rsa.PrivateKey
	TokenTTL  time.Duration
	TokenMax  int
	Now       func() time.Time

	mu     sync.Mutex
	tokens map[ghTokenKey]ghToken
	order  []ghTokenKey
}

type ghTokenKey struct {
	InstallationID int64
	AppID          int64
}

type ghToken struct {
	value     string
	expiresAt time.Time
}

// NewGHClient wires a GHClient from a Config.
func NewGHClient(cfg Config) (*GHClient, error) {
	orgKey, err := jwt.ParseRSAPrivateKeyFromPEM([]byte(cfg.GHAppOrgKey))
	if err != nil {
		return nil, fmt.Errorf("parse GHAPP_ORG_PRIVATE_KEY: %w", err)
	}
	persKey, err := jwt.ParseRSAPrivateKeyFromPEM([]byte(cfg.GHAppPersonalKey))
	if err != nil {
		return nil, fmt.Errorf("parse GHAPP_PERSONAL_PRIVATE_KEY: %w", err)
	}
	return &GHClient{
		HTTP:      http.DefaultClient,
		BaseURL:   "https://api.github.com",
		OrgAppID:  GHAppOrgID,
		PersAppID: GHAppPersonalID,
		OrgKey:    orgKey,
		PersKey:   persKey,
		TokenTTL:  59 * time.Minute,
		TokenMax:  1024,
		Now:       time.Now,
		tokens:    map[ghTokenKey]ghToken{},
	}, nil
}

func (c *GHClient) now() time.Time {
	if c.Now != nil {
		return c.Now()
	}
	return time.Now()
}

// GenerateJWT signs the app JWT with the matching private key.
// Exported so tests can verify produced tokens.
func (c *GHClient) GenerateJWT(appID int64) (string, error) {
	var key *rsa.PrivateKey
	switch appID {
	case c.OrgAppID:
		key = c.OrgKey
	case c.PersAppID:
		key = c.PersKey
	default:
		return "", fmt.Errorf("unknown app_id: %d", appID)
	}
	now := c.now()
	claims := jwt.MapClaims{
		"iat": now.Unix(),
		"exp": now.Add(10 * time.Minute).Unix(),
		"iss": appID,
	}
	return jwt.NewWithClaims(jwt.SigningMethodRS256, claims).SignedString(key)
}

func (c *GHClient) AuthenticateApp(ctx context.Context, installationID, appID int64) (string, error) {
	key := ghTokenKey{installationID, appID}
	c.mu.Lock()
	if t, ok := c.tokens[key]; ok && c.now().Before(t.expiresAt) {
		c.mu.Unlock()
		return t.value, nil
	}
	c.mu.Unlock()

	jwtTok, err := c.GenerateJWT(appID)
	if err != nil {
		return "", err
	}
	body, status, err := c.doJSON(ctx, "POST", "/app/installations/"+i64(installationID)+"/access_tokens", nil, jwtTok)
	if err != nil {
		return "", err
	}
	if status != 201 {
		return "", &GitHubAPIError{StatusCode: status, Message: apiMessage(body)}
	}
	var resp struct {
		Token string `json:"token"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return "", err
	}

	c.mu.Lock()
	c.tokens[key] = ghToken{value: resp.Token, expiresAt: c.now().Add(c.TokenTTL)}
	c.order = append(c.order, key)
	for len(c.order) > c.TokenMax {
		delete(c.tokens, c.order[0])
		c.order = c.order[1:]
	}
	c.mu.Unlock()
	return resp.Token, nil
}

func (c *GHClient) GetInstallation(ctx context.Context, installationID int64, et EntityType) (Installation, error) {
	appID := c.OrgAppID
	if et == EntityUser {
		appID = c.PersAppID
	}
	jwtTok, err := c.GenerateJWT(appID)
	if err != nil {
		return Installation{}, err
	}
	body, status, err := c.doJSON(ctx, "GET", "/app/installations/"+i64(installationID), nil, jwtTok)
	if err != nil {
		return Installation{}, err
	}
	if status != 200 {
		return Installation{}, &GitHubAPIError{StatusCode: status, Message: apiMessage(body)}
	}
	var inst Installation
	if err := json.Unmarshal(body, &inst); err != nil {
		return Installation{}, err
	}
	inst.Raw = body
	return inst, nil
}

// EnsureRunnerGroup returns the existing group id or creates the group.
// Callers log success with their full Entity context.
func (c *GHClient) EnsureRunnerGroup(ctx context.Context, token, orgName, groupName string) (int64, error) {
	listURL := "/orgs/" + orgName + "/actions/runner-groups"
	body, status, err := c.doJSON(ctx, "GET", listURL, nil, token)
	if err != nil {
		return 0, err
	}
	if status != 200 {
		return 0, &GitHubAPIError{StatusCode: status, Message: apiMessage(body)}
	}
	var lst struct {
		RunnerGroups []struct {
			ID   int64  `json:"id"`
			Name string `json:"name"`
		} `json:"runner_groups"`
	}
	if err := json.Unmarshal(body, &lst); err != nil {
		return 0, err
	}
	for _, g := range lst.RunnerGroups {
		if g.Name == groupName {
			return g.ID, nil
		}
	}
	createBody := map[string]any{
		"name":                       groupName,
		"visibility":                 "all",
		"allows_public_repositories": true,
	}
	body, status, err = c.doJSON(ctx, "POST", listURL, createBody, token)
	if err != nil {
		return 0, err
	}
	if status != 201 {
		return 0, &GitHubAPIError{StatusCode: status, Message: apiMessage(body)}
	}
	var created struct {
		ID int64 `json:"id"`
	}
	if err := json.Unmarshal(body, &created); err != nil {
		return 0, err
	}
	return created.ID, nil
}

func (c *GHClient) CreateJITRunnerConfigOrg(ctx context.Context, token, orgName, runnerName string, groupID int64, labels []string) (string, error) {
	body := map[string]any{
		"name":            runnerName,
		"runner_group_id": groupID,
		"labels":          labels,
		"work_folder":     "../../../work",
	}
	return c.jitconfig(ctx, "/orgs/"+orgName+"/actions/runners/generate-jitconfig", body, token)
}

func (c *GHClient) CreateJITRunnerConfigRepo(ctx context.Context, token, repoFullName, runnerName string, labels []string) (string, error) {
	body := map[string]any{
		"name":            runnerName,
		"runner_group_id": 1,
		"labels":          labels,
		"work_folder":     "../../../work",
	}
	return c.jitconfig(ctx, "/repos/"+repoFullName+"/actions/runners/generate-jitconfig", body, token)
}

func (c *GHClient) jitconfig(ctx context.Context, path string, body map[string]any, token string) (string, error) {
	respBody, status, err := c.doJSON(ctx, "POST", path, body, token)
	if err != nil {
		return "", err
	}
	if status != 201 {
		return "", &GitHubAPIError{StatusCode: status, Message: apiMessage(respBody)}
	}
	var resp struct {
		EncodedJITConfig string `json:"encoded_jit_config"`
	}
	if err := json.Unmarshal(respBody, &resp); err != nil {
		return "", err
	}
	return resp.EncodedJITConfig, nil
}

func (c *GHClient) ListRunnersOrgGroup(ctx context.Context, token, orgName string, groupID int64) ([]GHRunner, error) {
	return c.paginatedRunners(ctx, "/orgs/"+orgName+"/actions/runner-groups/"+i64(groupID)+"/runners", token)
}

func (c *GHClient) ListRunnersRepo(ctx context.Context, token, repoFullName string) ([]GHRunner, error) {
	return c.paginatedRunners(ctx, "/repos/"+repoFullName+"/actions/runners", token)
}

func (c *GHClient) paginatedRunners(ctx context.Context, path, token string) ([]GHRunner, error) {
	var all []GHRunner
	next := c.BaseURL + path + "?per_page=100"
	for next != "" {
		req, _ := http.NewRequestWithContext(ctx, "GET", next, nil)
		req.Header.Set("Authorization", "Bearer "+token)
		req.Header.Set("Accept", "application/vnd.github.v3+json")
		resp, err := c.HTTP.Do(req)
		if err != nil {
			return nil, err
		}
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		if resp.StatusCode != 200 {
			return nil, &GitHubAPIError{StatusCode: resp.StatusCode, Message: fmt.Sprintf("GET %s: %s", next, body)}
		}
		var page struct {
			Runners []GHRunner `json:"runners"`
		}
		if err := json.Unmarshal(body, &page); err != nil {
			return nil, err
		}
		all = append(all, page.Runners...)
		next = nextLink(resp.Header.Get("Link"))
	}
	return all, nil
}

// nextLink returns the URL for rel="next" in a GitHub Link header, or "".
func nextLink(header string) string {
	for _, part := range strings.Split(header, ",") {
		p := strings.TrimSpace(part)
		if !strings.HasSuffix(p, `rel="next"`) {
			continue
		}
		lt := strings.Index(p, "<")
		gt := strings.Index(p, ">")
		if lt >= 0 && gt > lt {
			return p[lt+1 : gt]
		}
	}
	return ""
}

func (c *GHClient) DeleteRunnerOrg(ctx context.Context, token, orgName string, runnerID int64) error {
	return c.deleteRunner(ctx, "/orgs/"+orgName+"/actions/runners/"+i64(runnerID), token)
}

func (c *GHClient) DeleteRunnerRepo(ctx context.Context, token, repoFullName string, runnerID int64) error {
	return c.deleteRunner(ctx, "/repos/"+repoFullName+"/actions/runners/"+i64(runnerID), token)
}

func (c *GHClient) deleteRunner(ctx context.Context, path, token string) error {
	body, status, err := c.doJSON(ctx, "DELETE", path, nil, token)
	if err != nil {
		return err
	}
	if status == 204 || status == 404 {
		return nil
	}
	return &GitHubAPIError{StatusCode: status, Message: fmt.Sprintf("DELETE %s: %s", path, body)}
}

func (c *GHClient) GetJobInfo(ctx context.Context, token, repoFullName string, jobID int64) (GHJob, error) {
	body, status, err := c.doJSON(ctx, "GET", "/repos/"+repoFullName+"/actions/jobs/"+i64(jobID), nil, token)
	if err != nil {
		return GHJob{}, err
	}
	if status != 200 {
		return GHJob{}, &GitHubAPIError{StatusCode: status, Message: fmt.Sprintf("get job %d: %s", jobID, body)}
	}
	var j GHJob
	if err := json.Unmarshal(body, &j); err != nil {
		return GHJob{}, err
	}
	return j, nil
}

func (c *GHClient) GetRunInfo(ctx context.Context, token, repoFullName string, runID int64) (GHRun, error) {
	body, status, err := c.doJSON(ctx, "GET", "/repos/"+repoFullName+"/actions/runs/"+i64(runID), nil, token)
	if err != nil {
		return GHRun{}, err
	}
	if status != 200 {
		return GHRun{}, &GitHubAPIError{StatusCode: status, Message: fmt.Sprintf("get run %d: %s", runID, body)}
	}
	var r GHRun
	if err := json.Unmarshal(body, &r); err != nil {
		return GHRun{}, err
	}
	return r, nil
}

// doJSON sends a JSON-encoded request and returns body, status, err.
// `auth` is the value put after "Bearer " in the Authorization header.
func (c *GHClient) doJSON(ctx context.Context, method, path string, body any, auth string) ([]byte, int, error) {
	var buf io.Reader
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return nil, 0, err
		}
		buf = bytes.NewReader(b)
	}
	url := c.BaseURL + path
	req, err := http.NewRequestWithContext(ctx, method, url, buf)
	if err != nil {
		return nil, 0, err
	}
	req.Header.Set("Authorization", "Bearer "+auth)
	req.Header.Set("Accept", "application/vnd.github.v3+json")
	if buf != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	resp, err := c.HTTP.Do(req)
	if err != nil {
		return nil, 0, err
	}
	defer resp.Body.Close()
	rb, err := io.ReadAll(resp.Body)
	return rb, resp.StatusCode, err
}

// apiMessage best-effort extracts {"message":...} from a GitHub error body.
func apiMessage(body []byte) string {
	var m struct {
		Message string `json:"message"`
	}
	if json.Unmarshal(body, &m) == nil && m.Message != "" {
		return m.Message
	}
	return string(body)
}

func i64(n int64) string { return fmt.Sprintf("%d", n) }

var _ = errors.New // reserved
